File: src/com/dugonggames/elclasico/DigitsMain.java
package com.dugonggames.elclasico;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

public class DigitsMain {

    public static void main(String args[]) throws IOException {
        byte[] trainingimagesbytes = Files.readAllBytes(Paths.get("C:\\Users\\alexf\\Downloads\\train-images-idx3-ubyte\\train-images.idx3-ubyte"));
        byte[] traininglabelsbytes = Files.readAllBytes(Paths.get("C:\\Users\\alexf\\Downloads\\train-labels-idx1-ubyte\\train-labels.idx1-ubyte"));
        byte[] testimagesbytes = Files.readAllBytes(Paths.get("C:\\Users\\alexf\\Downloads\\t10k-images-idx3-ubyte\\t10k-images.idx3-ubyte"));
        byte[] testlabelsbytes = Files.readAllBytes(Paths.get("C:\\Users\\alexf\\Downloads\\t10k-labels-idx1-ubyte\\t10k-labels.idx1-ubyte"));

        DigitImage[] trainingimages = new DigitImage[60000];
        for (int i = 0; i < 60000; i++) trainingimages[i] = new DigitImage(trainingimagesbytes, 16+(784*i), traininglabelsbytes, 8+i);

        Tree t = new Tree(trainingimages);

        DigitImage[] testimages = new DigitImage[60000];
        for (int i = 0; i < 10000; i++) testimages[i] = new DigitImage(testimagesbytes, 16+(784*i), testlabelsbytes, 8+i);

        int correct = 0;
        for (int i = 0; i < 10000; i++){
           if (testimages[i].getLabel() == t.classify(testimages[i])) correct++;
        }
        System.out.println(correct);

        int num = 543;
        testimages[num].print();
        System.out.println(t.classify(testimages[num]));
        System.out.println(testimages[num].getLabel());
    }
}

File: src/main/kotlin/com/dugonggames/elclasico/DigitsMain.kt
package com.dugonggames.elclasico

import com.dugonggames.elclasico.Classifiers.BuildTree.buildForest
import com.dugonggames.elclasico.Classifiers.BuildTree.buildTree
import com.dugonggames.elclasico.Classifiers.FeatureVector
import com.dugonggames.elclasico.Classifiers.LabeledSample
import com.dugonggames.elclasico.ImageManipulation.Pyramid
import com.dugonggames.elclasico.ImageManipulation.createGaussianPyramid
import com.dugonggames.elclasico.ImageManipulation.createLaplacianPyramid
import java.nio.file.Paths

object DigitsMain {

    @JvmStatic
    fun main(args: Array<String>) {
        val root = System.getProperty("user.home")
        val rootDataDirectory = Paths.get(root).resolve("Downloads").resolve("dataset")
        val trainingimagesbytes = rootDataDirectory.resolve("train-images.idx3-ubyte").toFile().readBytes()
        val traininglabelsbytes = rootDataDirectory.resolve("train-labels.idx1-ubyte").toFile().readBytes()
        val testimagesbytes = rootDataDirectory.resolve("t10k-images.idx3-ubyte").toFile().readBytes()
        val testlabelsbytes = rootDataDirectory.resolve("t10k-labels.idx1-ubyte").toFile().readBytes()

        println("Loading images.")
        val trainingimages = Array(60000){ i ->
            val d=DigitImage.fromFile(trainingimagesbytes, 16 + 784 * i, traininglabelsbytes, 8 + i)
            LabeledSample(preprocessImage(d), d.label)
        }
        println("Done loading.")

        //trainingimages.forEach { println(it.fv[387]) }

        //val t = buildTree(images = trainingimages, numClasses = 10, maxDepth = 1)

        val f = buildForest(20, trainingimages, 10, 6, 28, 20)

        val testimages = Array(10000) {
            i -> DigitImage.fromFile(testimagesbytes, 16 + 784 * i, testlabelsbytes, 8 + i)
        }

        val correct = testimages.count { it.label == f.classify(preprocessImage(it)) }
        println(correct)
    }

    fun preprocessImage(digitImage: DigitImage): FeatureVector {
        return FeatureVector(digitImage.digit)
        //return FeatureVector(createGaussianPyramid(digitImage.digit, 28, 28).data)
        //return FeatureVector(createLaplacianPyramid(digitImage.digit, 28, 28).data)
    }
}
